ncrypt-0.6.11/ncrypt.8
.TH NCRYPT 8 "August 2004" NMRC "User Manuals"

.SH NAME

ncrypt \- NMRC File Encryptor/Decryptor/Wiper

.SH SYNOPSIS

.B ncrypt <-h> <-e|-d> <-z> <-w> <-v> <-V> <-k
.I pass
.B |-f 
.I passfile
.B > <-a
.I algo
.B > <-i
.I infile
.B > <-o
.I outfile
.B >

.SH DESCRIPTION

Ncrypt is intended to give you security in an insecure environment. If you
are wanting to encrypt files (particularly on a multi-user system where you
don't have root), wishing to hide your activites from prying eyes, and want
to "cover your tracks", then Ncrypt is for you. It gives you the choice of
three strong encryption algorithms and two secure file wiping options.

For file encryption, Ncrypt is a symmetrical file encryptor/decryptor that 
gives you the choice of the top three candidates for AES as the encryption
algorithm (Rijndael (AES), Serpent, Twofish), tries to minimize exposure of 
the plaintext password in memory, and can safely erase the plaintext version 
from the hard drive. 

The erasing of files uses techniques from two sources. The first set of
techniques was originally outlined in Peter Gutmann's 1996 paper "Secure 
Deletion of Data from Magnetic and Solid-State Memory". The second set of
techniques was originally outlined in the DoD 5220.22-M standard developed
by the NSA, specifically the recommendations for sanitizing drives for
redeployment. For the insanely paranoid, yes you can use both options at the
same time.
 
For added security during wiping, the random data streams use the ISAAC PRNG 
during wiping of files, and the wiping can take place either independently 
or concurrently with encryption operations. The wiping routine also wipes 
file slack. 

Ncrypt compiles without any extra crypto libraries, making it ideal for 
systems where you just have a compiler and basic libraries (such as an ISP's 
shell server).

.SH OPTIONS
.IP "-h or --help"
Display help/usage information and exit.  
.IP "-e or --encrypt"
Signifies you wish to encrypt a file. When encrypting you are asked for the
password twice (to ensure you have entered it correctly).
.IP "-d or --decrypt"
Signifies you wish to decrypt a file. When decrypting you are asked for the
password one time.
.IP "-r or --rijndael or -a or --aes"
Use the AES (Rijndael) algorithm for encryption/decryption. If no algorithm is 
chosen, Rijndael is used by default.
.IP "-s or --serpent"
Use the Serpent algorithm for encryption/decryption.
.IP "-t or --twofish"
Use the Twofish algorithm for encryption/decryption.  
.IP "-i or --input"
Specify the input file \fIinfile\fP for ncrypt to either encrypt or decrypt. 
If used with the -g or -m option during encryption, the input file specified 
is wiped.
.IP "-o or --output"
Specify the output file \fIoutfile\fP for the results of ncrypt's encryption
or decryption.
.IP "-z or --encrypt-wipe"
During encryption, wipe the input file after it has been encrypted using
either the Gutmann or military method. If neither are chosen, military mode
is used by default. Careful - the erased file is unrecoverable, which is
kind of the point.
.IP "-g or --wipe-gutmann"
Do a `safe' erase of the input file (during encryption mode) or a specified
file or files if not performing encryption/decryption operations. This
overwrites the file on the hard drive with a series of 35 passes as outlined
in Peter Gutmann's 1996 paper (see above) before erasing the file. File slack
is also wiped using the same technique.
.IP "-m or --wipe-military"
This option runs like the -g or --wipe-gutmann option, except 3 passes are
done. The first pass is a character, the second is the complement of that
character (the bits are flipped), and the third is with random data which is
verified. This is the military-grade standard for wiping files that are
classified (does not apply to TOP SECRET classification). And like the -g or 
--wipe-gutmann, use with caution as the file is not recoverable.
.IP "-k or --key passphrase"
This is a somewhat dangerous option, since you are using the passphrase on the
command line, but it does allow you to include Ncrypt in batch files or cron
jobs. Comes in handy for encrypting archived logs before using an insecure 
plaintext method of copying them to another system, such as email. Just
remember that the ps command could reveal the passphrase to anyone on the
system. This option can be used for encryption or decryption.
.IP "-f or --password-file filename.txt"
Also a somewhat dangerous command, reads the passphrase from a file. The file
must be a plaintext file, and the first line in the file that doesn't start
with a pound (#) sign for comments is used as the passphrase. The ending
carriage return is stripped off, and you can use up to a 255 character 
passphrase. The file containing the password should be readable by the UID/EUID
of the person or process running Ncrypt. Like the -k option, it comes in handy 
for batch files and cron jobs, and it can be used for encryption or decryption.
.IP "-v or --verbose"
More verbose output during program execution.
.IP "-V or --version"
Print ncrypt version information and exit.

.SH EXAMPLE USAGE
.IP "\fBncrypt -e -s -i foo.txt -o bar.enc\fR"
Encrypts plaintext file foo.txt and writes the encrypted file out as bar.enc.
The Serpent algorithm is being used.
.IP "\fBncrypt -d -i bar.enc -o foo.txt\fR"
Decrypts encrypted file bar.enc and writes out plaintext file foo.txt. Since
no algorithm is specified, AES (Rijndael) is used by default.
.IP "\fBncrypt --encrypt --wipe-gutmann -i foo.txt -o bar.enc\fR"
Encrypts plaintext file foo.txt and writes the encrypted file out as bar.enc.
AES is being used, and the original foo.txt file using the Gutmann method.
.IP "\fBncrypt -m foo.txt\fR"
The file foo.txt is erased using the military method.
.IP "\fBncrypt --wipe-gutmann foo.txt bar.txt foobar.txt\fR"
The files foo.txt, bar.txt, and foobar.txt are wiped using the Gutmann method.
.IP "\fBncrypt -e -z -k secretpassphrase -i foo.txt -o bar.enc\fR"
The file foo.txt is encrypted with the password of "secretpassphrase" into
bar.enc, and then foo.txt is securely wiped via the -z option with the default
military method.

.SH DIAGNOSTICS
Errors from Ncrypt are written to stderr. There shouldn't be any bugs, because
it compiled without errors ;-)

.SH AUTHOR
Ncrypt implementation by Simple Nomad <thegnome@nmrc.org>. Extensive code
review and additional coding by Inertia <inertia@nmrc.org>. Additional coding 
ideas, including the improved PRNG were added by Todd MacDermid 
<tmacd@synacklabs.net>. Based upon code by Joh Johnson <jojo@farm9.com>, 
which was taken from code by Gary Rancier, as well as code taken from Dave 
Whiting's Twofish implementation. The read_pass.c routine is based upon 
sample code in the excellent book "Building Secure Software" by John Viega
and Gary McGraw.

.SH BUGS
Send bug reports to <thegnome@nmrc.org>.


ncrypt-0.6.11/NEWS
HISTORY
-------

$Id: NEWS,v 1.7 2004/03/17 05:02:56 s-nomad Exp $

0.6.9
16Mar2004 - Code is pretty ugly, hard to troubleshoot the post arg processing
            so simplified things. Removed all of Inertia's code and redid
            the entire getopt processing. There was a problem that if someone
            forgot to specify the -e, ncrypt would interpret that as a wipe
            only operation and securely wipe the input file. This is fixed.
            General code cleanup on ncrypt.c, removed cmdline.c and cmdline.h
            entirely. Will probably go through and clean up remaining code.

0.6.8
15Sep2003 - Naming conflict with XPG4.2 within unistd.h, renamed a routine.

0.6.7
09Sep2003 - Ugh. Dependency problems. Redid configure.in from scratch, and
            redid all of the automake support files. Stuff seems to build 
            and install ok now without too many errors. Edited cmdline.h
            to eliminate non-printing of package and version info. Tweaked
            the man page. Minor tweaks to rand_gen.c, cmdline.c, and
            get_encryption_password.c to eliminate compilation errors.

0.6.6
16Aug2003 - Updated configure.ac so that it will work with autoconf 2.53.
            Added 'debian' directory to version control.

0.6.5
03Aug2003 - Fixed minor typo in help info. Adjusted some logic in the main
            program, as illogical parameter choices caused illogical things
            to happen. Set default wipe mode to military if not specified.

0.6.4
14Jul2003 - NTFS and djgpp don't seem to get along when it comes to handling
            the file slack wiping and the verify in military wiping, so
            currently the WIN32 version does not support it. Completed the
            WIN32 version, fixed minor bug with --aes option in cmdline.c.

09May2003 - Added routines for doing DoD 5220.22-M style of wiping, giving the
            choice of two wiping methods. The DoD method uses 3 passes, the
            third pass being verified as it is being to disk. Added a -a
            (--aes) option which simply points to Rijndael. Updated all of
            the relevant files (README, man page, etc). Fixed a bug in the
            install routine which tried to install the header files in this
            distro.

03May2003 - Added better error handling to wipe_file.c, and added routine to
            perform wiping on file slack of file being wiped.

01May2003 - fopen truncates the file before returning, so updated wipe_file
            to use open instead. Wiping is much slower now, so added UI
            indications in --verbose mode that something is actually
            happening. Updated man page to make it more accurate.

0.6.3
26Apr2003 - Oops, odd logic bug caused ncrypt to never encrypt or decrypt
            a file, only show the help screen, ugh. Adjusted read_pass to
            use tcsetattr and tcgetattr, making the code more portable.
            Minor tweaks in ncrypt.c and rand_gen.c so it will compile
            under djgpp. Created a makefile for djgpp. Added local getopt
            routines if target does not have getopt or getopt_long (Mac 
            OS X, djgpp).

0.6.2
23Apr2003 - Two bugs in shuffling of order of the constant overwrites in
            wipe_file.c, updated the README to reflect the current version.

0.6.1
22Apr2003 - Autoconf and automake. Use of long options as well. A lot of
            code cleanup. Added, then removed I/O using stdin/stdout instead
            of files (too many security issues as of yet).

21Apr2003 - Switched the random number seeding algorithm used during
            file deletion to one based on processor timing events, and
            the PRNG from rand() to ISAAC. Also shuffled the order
            of the constant overwrites, as recommended by Peter
            Gutmann.

0.5.5
14Feb2003 - Fixed a bug that allowed -w to work during encryption, which
            completely erased the input file and there was no output file
            at all (oops). Applied patch so read_pass compiles and runs on
            FreeBSD, and adjusted how memory is locked to prevent paging
            (not all systems support mlockall).

0.5.4
30Jan2003 - Updated the wipe_file.c program to use the wiping techniques
            discussed in "Secure Deletion of Data from Magnetic and Solid-State
            Memory" by Peter Gutmann. The original paper can be found at
            http://www.cs.auckland.ac.nz/~pgut001/pubs/secure_del.html. This
            changes the wiping algo for a total of 35 overwrites. Updated
            the README as well.
31Dec2002 - Added mem.c routine so that if optimized, we don't rely on memset
            but use our own. Updated the Makefile. Added a TODO file. Added
            code to check for running as root, and if so try to mlockall to
            prevent paging memory to disk. Updated the README.

0.5.3       ** non-release **

0.5.2
19Dec2002 - Fixed minor bug in wipe_file routine. Minor code cleanup and updates
            to the README and man page.
17Dec2002 - Moved some random number stuff to rand_gen.c. Moved the wipe_file
            routine to wipefile.c. Increased the overwrites to a total of
            27 times (14 passes with zeroes, 13 passes with random material). 
            Added -V switch to print version. Updated the man page.
18Sep2002 - Added string.h to readpass.c to eliminate an error. Added a
            HISTORY file.

0.5.1
09Sep2002 - Added README and built man page.
05Sep2002 - Code cleanup, tweaks, and minor bugfixes to Twofish code. Added
            WIPEONLY mode to do the wiping on a single file.
04Sep2002 - Added Twofish, had to remove the key block size.

0.4.1
03Sep2002 - Adjusted key block size to a variable in ncrypt.h. Added install
            routine to Makefile.
31Aug2002 - Minor code cleanup, added code to wipe hash from memory after
            it is no longer needed.

0.4

ncrypt-0.6.11/ncrypt.c
/* 
 * ncrypt.c - Simple Nomad <thegnome@nmrc.org>
 *          - Inertia <inertia@nmrc.org>
 *
 * NMRC file encryptor/decryptor - based off of AES code for Rijndael and
 * Serpent found on www.farm9.com. The original farm9 work done by Joh 
 * Johannsen <jojo@farm9.com>, which was based off of work done by Gary 
 * Rancier <mephis5@softhome.net>. The Twofish code was taken from Doug 
 * Whiting's C implementation from www.counterpane.com.
 * 
 * $Id: ncrypt.c,v 1.10 2004/08/25 20:20:36 s-nomad Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <unistd.h>
#include "ncrypt.h"
#include "config.h"
#include "encrypt_file.h"
#include "decrypt_file.h"
#include "wipe_file.h"
#include "rand_gen.h"
#include "read_pass.h"
#include "sha1.h"
#include "mem.h"
#include "get_encryption_password.h"
#include "get_decryption_password.h"

#ifdef __FREEBSD__
#define LOCKING 0
#else
#define LOCKING 1
#endif

/**********
 * globals
 **********/
char* cli_key;
char* fpass;
keyInstanceT keyInstT;
keyInstanceS keyInstS;
keyInstanceR keyInstR;
cipherInstance cipherInst;
cipherInstanceT cipherInstT;
struct ncrypt NCRYPT;
const char* PROGNAME;

/**********
 * routines
 **********/

void initialize_options( void )
{
  NCRYPT.verbose = 0;
  NCRYPT.algo = 0;
  NCRYPT.key = 0;
  NCRYPT.pfile = 0;
}

void initialize_crypto(char* pass, int encrypt)
{
    BYTE keyMaterial[320];
    BYTE direction = (encrypt) ? DIR_ENCRYPT : DIR_DECRYPT;
    int r = 0;
    int i = 0;
    int len = 256; 

    if(NCRYPT.algo == TWOFISH) len = 128; 

    /* make plaintext password into a SHA1 hash */
    extend_mat(hash_string_with_sha1(pass), &keyMaterial[0], len);

    /* after hashing, erase the plaintext password instance from memory */
    guaranteed_memset(pass,0,strlen(pass));

    switch(NCRYPT.algo)
    {
      case RIJNDAEL:
        r = makeKeyR(&keyInstR, direction, len, keyMaterial);
        break;
      case SERPENT:
        r = makeKeyS(&keyInstS, direction, len, keyMaterial);
        break;
      case TWOFISH:
        r = makeKeyT(&keyInstT, direction, len, NULL);
        if (r == TRUE)
        {
            for(i=0; i<(len/32); i++) keyInstT.key32[i] = keyMaterial[i];
            reKey(&keyInstT);
        }
        break;
    }

    /* we've made our key from the hash, so remove the hash from memory */
    guaranteed_memset(keyMaterial,0,len);

    if (r != TRUE)
    {
        fprintf(stderr,"** ERROR: On makeKey: %d\n", r);
        exit(-1);
    }

    /* init the algo we are using */
    switch(NCRYPT.algo)
    {
      case RIJNDAEL:
        r = cipherInitR(&cipherInst,  MODE_CBC, NULL); break;
      case SERPENT:
        r = cipherInitS(&cipherInst,  MODE_ECB, NULL); break;
      case TWOFISH:
        r = cipherInitT(&cipherInstT, MODE_CBC, NULL); break;
    }
    if (r != TRUE)
    {
        fprintf(stderr,"** ERROR: On cipherInit: %d\n", r);
        exit(-1);
    }
}


#ifndef WIN32
void check_for_lockable_memory()
{
    int lock = 0;
    if (getuid() == 0)
    {
        lock = mlockall(MCL_CURRENT|MCL_FUTURE);
        if (lock == -1)
        {
            fprintf(stderr,"Running as root but unable to lock memory to prevent paging to swap\n");
            exit(-1);
        }
        lock = 1;
    }
    if (NCRYPT.verbose)
    {
	if (lock)
            fprintf(stderr,"NOTE: Running as root and memory is locked from paging to disk\n");
        else
            fprintf(stderr,"WARNING: Running unprivileged and memory is not locked from paging to disk\n");
    }
}
#endif


int wipe(int wipe_mode, char * ifile)
{
    int i = 0;
    
    if(ifile)
	    wipe_file(ifile,NCRYPT.verbose,wipe_mode);
    else for (i = 0 ; i < NCRYPT.ifiles_cnt ; i++)
	wipe_file(NCRYPT.ifiles[i],NCRYPT.verbose,wipe_mode);
    return 0;
}

int encrypt_it(FILE* ifp, FILE* ofp)
{
    struct stat statbuf;
    char* pass = NULL;

    if(NCRYPT.key)
    {
        initialize_crypto(cli_key,1);
        guaranteed_memset(cli_key,0,strlen(cli_key));
        free(cli_key);
    }
    else if(NCRYPT.pfile)
    {
        initialize_crypto(fpass,1);
        guaranteed_memset(fpass,0,strlen(fpass));
        free(fpass);
    }
    else
    {
        pass = get_encryption_password();
        initialize_crypto(pass,1);
        guaranteed_memset(pass,0,strlen(pass));
        free(pass);
    }
    fstat(fileno(ifp),&statbuf);
    encrypt_file(ifp,statbuf.st_size,ofp,NCRYPT.algo);
    return 0;
}



int decrypt_it(FILE* ifp, FILE* ofp)
{
    char* pass = NULL;
    if(NCRYPT.key)
    {
        initialize_crypto(cli_key,0);
        guaranteed_memset(cli_key,0,strlen(cli_key));
        free(cli_key);
    }
    else if(NCRYPT.pfile)
    {
        initialize_crypto(fpass,0);
        guaranteed_memset(fpass,0,strlen(fpass));
        free(fpass);
    }
    else
    {
        pass = get_decryption_password();
        initialize_crypto(pass,0);
        guaranteed_memset(pass,0,strlen(pass));
        free(pass);
    }
    decrypt_file(ifp,ofp,NCRYPT.algo);
    return 0;
}



FILE* determine_input_file(char * ifile)
{
    FILE* ifp;
    ifp = fopen(ifile,"rb");
    if (ifp == NULL)
    {
      fprintf(stderr,"ERROR: cannot open file for reading: %s\n",ifile);
      exit(-1); 
    }
    return ifp;
}



FILE* determine_output_file(char * ofile)
{
    FILE* ofp;
    ofp = fopen(ofile,"wb");
    if (ofp == NULL)
    {
        fprintf(stderr,"ERROR: cannot open file for writing: %s\n",ofile);
        exit(-1); 
    }
    return ofp;
}

#ifndef HAVE_STRDUP
/* gengetopt_strdup(): automatically generated from strdup.c. */
/* strdup.c replacement of strdup, which is not standard */
static char *gengetopt_strdup (const char *s)
{
        char *result = (char*)malloc(strlen(s) + 1);
        if (result == (char*)0)
                return (char*)0;
        strcpy(result, s);
        return result;
}
#endif /* HAVE_STRDUP */

void usage(int exit_state)
{
    printf("USAGE:\n"
    "  ncrypt <options> [ops] <-i in_file> <-o out_file> <wipe-only files>\n"
    "  Options include the following:\n"
    "    -h  --help          This help cruft\n"
    "    -V  --version       Print version and exit\n"
    "    -v  --verbose       Verbose mode\n"
    "    -a  --aes           Use AES when encrypting/decrypting (same as rijndael)\n"
    "    -r  --rijndael      Use rijndael when encrypting/decrypting (default)\n"
    "    -s  --serpent       Use serpent when encrypting/decrypting\n"
    "    -t  --twofish       Use twofish when encrypting/decrypting\n"
    "    -z  --encrypt-wipe  Wipe input file after encrypting\n"
    "    -k  --key <pass>    Supply the password on the commandline (dangerous)\n"
    "    -f  --password-file <filename.txt>\n"
    "                        Supply the password in a file (also dangerous)\n"
    "  Ops include the following:\n"
    "    -e  --encrypt       Encryption mode\n"
    "    -d  --decrypt       Decryption mode\n"
    "    -g  --wipe-gutmann  Wipe using Gutmann mode\n"
    "    -m  --wipe-military Wipe using DoD 5220.22-M (military grade) mode\n"
    "  More info, including usage examples, in the man page\n\n");
    exit(exit_state);
}

int main(int argc, char** argv)
{
    FILE* ifp = NULL;
    FILE* ofp = NULL;
    FILE* pfp = NULL;
    char * ifile = NULL;
    char * ofile = NULL;
    char * pfile = NULL;
    char buf[MAX_LINE+1];
    int encrypt=0, decrypt=0, line=0;
    int wipe_gutmann=0, wipe_military=0, wipe_only=0;
    int rc = 0, wipe_mode = 0, wipe_on_encrypt=0;
    int c,i,algo=0,wiping=0;
    int option_index = 0;
    static struct option long_options[] = {
      { "help",       0, NULL, 'h' },
      { "version",    0, NULL, 'V' },
      { "input",      1, NULL, 'i' },
      { "output",     1, NULL, 'o' },
      { "key",        1, NULL, 'k' },
      { "password-file",      1, NULL, 'f' },
      { "wipe-gutmann",       0, NULL, 'g' },
      { "verbose",    0, NULL, 'v' },
      { "encrypt",    0, NULL, 'e' },
      { "decrypt",    0, NULL, 'd' },
      { "wipe-military",      0, NULL, 'm' },
      { "aes",        0, NULL, 'a' },
      { "rijndael",   0, NULL, 'r' },
      { "serpent",    0, NULL, 's' },
      { "twofish",    0, NULL, 't' },
      { "encrypt-wipe",       0, NULL, 'z' },
      { NULL, 0, NULL, 0 }
    };
    PROGNAME = argv[0];
    cli_key = NULL;

    initialize_options();
    optind = 1;

    while(1)
    {
      c = getopt_long(argc, argv, "hVi:o:k:f:gvedmarstz", long_options, &option_index);
      if(c == -1) break;
      switch(c)
      {
        case 'h':
          usage(EXIT_SUCCESS);
        case 'V':
          printf("%s %s\n", PACKAGE, VERSION);
          exit(EXIT_SUCCESS);
        case 'i':
          if(ifile)
          {
            fprintf(stderr,"%s: --input (-i) used more than once\n",PACKAGE);
            exit(EXIT_FAILURE);
          }
          ifile = strdup(optarg);
          break;
        case 'o':
          if(ofile)
          {
            fprintf(stderr,"%s: --output (-o) used more than once\n",PACKAGE);
            exit(EXIT_FAILURE);
          }
          ofile = strdup(optarg);
          break;
        case 'k':
          if(cli_key)
          {
            fprintf(stderr,"%s: --key (-k) used more than once\n", PACKAGE);
            exit(EXIT_FAILURE);
          }
          cli_key = strdup(optarg);
          NCRYPT.key = 1;
          break;
        case 'f':
          if(pfile)
          {
            fprintf(stderr,"%s: --password-file (-f) used more than once\n", PACKAGE);
            exit(EXIT_FAILURE);
          }
          pfile = strdup(optarg);
          pfp = fopen(pfile,"r");
          if(!pfp)
          {
            fprintf(stderr,"%s: unable to open file containing password\n", PACKAGE);
            exit(EXIT_FAILURE);
          }
          buf[MAX_LINE] = 0;
          while (!feof(pfp))
          {
            fgets(buf,MAX_LINE,pfp);
            line++;
            if(buf[0]=='#') continue;
            if(buf[0]!='#')
            {
              if(strlen(buf) < 1) continue;
              for(i=0;i<MAX_LINE;i++)
                if(buf[i]=='\n') buf[i] = 0;
              fpass = strdup(buf);
              break;
            }
          }
          fclose(pfp);
          if(!buf)
          {
            fprintf(stderr,"%s: unable to find a password in file %s\n", PACKAGE, pfile);
            exit(EXIT_FAILURE);
          }
          NCRYPT.pfile = 1;
          break;
        case 'g':
          wipe_gutmann = 1;
          wiping++;
          break;
        case 'v':
          NCRYPT.verbose = 1;
          break;
        case 'e':
          encrypt = 1;
          break;
        case 'd':
          decrypt = 1;
          break;
        case 'm':
          wipe_military = 1;
          wiping++;
          break;
        case 'a':
          algo++;
          NCRYPT.algo = RIJNDAEL;
          break;
        case 'r':
          NCRYPT.algo = RIJNDAEL;
          algo++;
          break;
        case 's':
          NCRYPT.algo = SERPENT;
          algo++;
          break;
        case 't':
          NCRYPT.algo = TWOFISH;
          algo++;
          break;
        case 'z':
          wipe_on_encrypt = 1;
          break;
        case 0:
        case '?':
          exit(EXIT_FAILURE);
        default:
          fprintf(stderr,"%s: option unknown: %c\n",PACKAGE,c);
          usage(EXIT_FAILURE);
      }
    }

    /* post arg processing, this is ugly, but at least it is all in one place */
    /* can't encrypt AND decrypt at the same time */
    if(encrypt && decrypt)
    {
      fprintf(stderr,"%s: You must choose either to encrypt or decrypt, not both\n",PACKAGE);
      usage(EXIT_FAILURE);
    }
    /* must pick something, if not encrypting or decrypting and haven't chosen
       a wiping method we assume nothing and error out accordingly */
    if(!encrypt && !decrypt && !wiping)
    {
      fprintf(stderr,"%s: You must encrypt, decrypt, or wipe\n",PACKAGE);
      usage(EXIT_FAILURE);
    }
    /* if decrypting we don't need to be wiping */
    if(decrypt && wiping)
    {
      fprintf(stderr,"%s: If decrypting, you shouldn't be choosing wiping options\n",PACKAGE);
      usage(EXIT_FAILURE);
    }    
    /* so if we aren't encrypting ot decrypting, we must be wiping */
    if(!encrypt && !decrypt && wiping) wipe_only = 1;
    /* however if someone used the -i or -o option, maybe they were trying
       to encrypt and left out the -e option, which could be bad since we
       could securely wipe something that wasn't intended to be wiped */
    if(wipe_only && ifile)
    {
      fprintf(stderr,"%s: If wiping only, do not use the -i option\n",PACKAGE);
      usage(EXIT_FAILURE);
    }
    if(wipe_only && ofile)
    {
      fprintf(stderr,"%s: If wiping only, do not use the -o option\n",PACKAGE);
      usage(EXIT_FAILURE);
    }
    /* if encrypting or decrypting, we need an input and output file */
    if(encrypt || decrypt)
    {
      if(!ifile)
      {
        fprintf (stderr, "%s: An input file is required for encryption or decryption\n", PACKAGE);
        usage(EXIT_FAILURE);
      }
      if(!ofile)
      {
        fprintf (stderr, "%s: An output file is required for encryption or decryption\n", PACKAGE);
        usage(EXIT_FAILURE);
      }
    }
    /* the algo should be chosen only once */
    if(algo > 1)
    {
      fprintf(stderr,"%s: You made algorithm choices more than once\n",PACKAGE);
      usage(EXIT_FAILURE);
    }
    /* can't do -k and -f together */
    if((NCRYPT.pfile) && (NCRYPT.key))
    {
      fprintf(stderr,"%s: You cannot use -k and -f together\n",PACKAGE);
      usage(EXIT_FAILURE);
    }
    /* if we reach here and we've processed all options, anything left on the
       commandline is copied into ifiles and we store the count */
    if(optind < argc)
    {
      int i = 0;
      NCRYPT.ifiles_cnt = argc - optind;
      NCRYPT.ifiles = (char **)(malloc ((NCRYPT.ifiles_cnt)*sizeof(char *)));
      while(optind < argc)
        NCRYPT.ifiles[i++] = strdup(argv[optind++]);
    }
    /* if no algo chosen, we default to rijndael */
    if (!NCRYPT.algo) NCRYPT.algo = RIJNDAEL;
    if (NCRYPT.verbose && ofile)
    {
      switch(NCRYPT.algo)
      {
        case RIJNDAEL: printf("NOTE: using aes (rijndael)\n"); break;
        case SERPENT: printf("NOTE: using serpent\n"); break;
        case TWOFISH: printf("NOTE: using twofish\n"); break;
      }
    }
#ifndef WIN32
    check_for_lockable_memory();
#endif
    /* if using files, verify we can actually use them and get the handles */
    if (ifile) ifp = determine_input_file(ifile);
    if (ofile) ofp = determine_output_file(ofile);
    
    /* if wiping, determine wiping mode */
    if (wipe_gutmann) wipe_mode = WIPE_GUTMANN;
    if (wipe_military) wipe_mode = WIPE_MILITARY;
    if ((wipe_gutmann) && (wipe_military)) wipe_mode = WIPE_BOTH;
    if (!wipe_mode) wipe_mode = WIPE_MILITARY; // z option with no mode chosen
    /* final end of post arg processing */

    /* finally doing something */
    if (wipe_only) return wipe(wipe_mode,NULL);

    if (encrypt) 
    {
	    rc = encrypt_it(ifp,ofp);
	    if (wipe_on_encrypt && (rc == 0)) rc = wipe(wipe_mode,ifile);
    }
    if (decrypt) rc = decrypt_it(ifp,ofp);

    fclose(ifp);
    fclose(ofp);
    return rc;
}
